### PR TITLE
DAOS-2448 obj: EC degraded fetch with data recovering

### PR DESCRIPTION
**Pull request by liuxuezhao**
Adds the support of recover the missed data on shadow epochs.

Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

### CODE REVIEW

File: src/client/api/object.c [hunk 1]
```diff
@@ -162,7 +162,7 @@ daos_obj_fetch(daos_handle_t oh, daos_handle_t th, uint64_t flags,
 	tse_task_t	*task;
 	int		rc;
 
-	rc = dc_obj_fetch_task_create(oh, th, flags, dkey, nr, iods, sgls,
+	rc = dc_obj_fetch_task_create(oh, th, flags, dkey, nr, iods, sgls, NULL,
```
Comment (wangdi1) [inline src/client/api/object.c:165]: Hmm, the extra arg has to be inside the task? can it be into obj_auxi_args? 
This is a general API, so it is probably not good to put extra special arg into the interface.

File: src/include/daos/object.h [hunk 2]
```diff
@@ -437,7 +439,7 @@ struct daos_recx_ep_list {
 	/** #total items (capacity) in re_items array */
 	uint32_t		 re_total;
 	/** epoch valid flag, re_items' re_ep can be ignored when it is false */
-	uint32_t		 re_ep_valid:1;
+	bool			 re_ep_valid;
```
Comment (wangdi1) [inline src/include/daos/object.h:442]: Just curious, why change this from 1 bit to 32 bit?

File: src/object/cli_ec.c [hunk 3]
```diff
@@ -999,6 +1024,53 @@ obj_reasb_req_dump(struct obj_reasb_req *reasb_req, d_sg_list_t *usgl,
...
+
+	buf_sgl = sgl->sg_iovs[0].iov_buf;
+	cell_sz = obj_ec_cell_rec_nr(oca) * iod_size;
+	stripe_total_sz = cell_sz * obj_ec_tgt_nr(oca);
+	stripe_nr = recx->rx_nr / stripe_rec_nr;
+	recx_nr = recx->rx_nr / obj_ec_data_tgt_nr(oca);
+	for (tgt = 0, tgt_nr = 0; tgt < obj_ec_tgt_nr(oca); tgt++) {
+		if (obj_ec_tgt_in_err(fail_info->efi_tgt_list,
+				      fail_info->efi_ntgts, tgt))
+			continue;
```
Comment (wangdi1) [inline src/object/cli_ec.c:1054]: so if it reach the maxim fault toleration, it does not need to recovery anymore?

File: src/object/cli_ec.c [hunk 4]
```diff
@@ -1502,6 +1663,434 @@ obj_ec_recov_prep(struct obj_reasb_req *reasb_req, daos_obj_id_t oid,
...
+}
+
+static void
+obj_ec_sgl_copy(d_sg_list_t *sgl, uint64_t off, void *buf, uint64_t size)
+{
+	struct daos_sgl_idx	sgl_idx = {0};
+	struct oes_copy_arg	arg;
+	int			rc;
+
+	rc = daos_sgl_processor(sgl, &sgl_idx, off, NULL, NULL);
```
Comment (wangdi1) [inline src/object/cli_ec.c:1977]: is this noop? Could you please add some comment here?

File: src/client/api/object.c (hunk 1 above)
Comment (liuxuezhao) [reply src/client/api/object.c:165]: at this time the task is to be created. probably can use tse_task_stack_push() to get the obj_auxi and set the extra arg pointer, and then pop it. that seems more complex, i can change to that way if you prefer it.
At my side, it seems not bad to add void *extra_arg to task args and task create API, and explains by different flags?

File: src/include/daos/object.h (hunk 2 above)
Comment (liuxuezhao) [reply src/include/daos/object.h:442]: since there is only one bit now, seems bool is more clear?

File: src/object/cli_ec.c (hunk 3 above)
Comment (liuxuezhao) [reply src/object/cli_ec.c:1054]: yes for now, and if #tgts_failed > limit, then it is not recoverable.
But probably some recoverable cases can be refined later, for example if some tgts in rebuilding, in that case maybe can return EAGAIN?

File: src/object/cli_ec.c (hunk 4 above)
Comment (liuxuezhao) [reply src/object/cli_ec.c:1977]: it is to skip sgl to off, i'll add some comments when refresh this patch, thx

**Comment by liuxuezhao** [comment]
will close and resubmit to master branch, as this one was submitted to other branch.

**Comment by liuxuezhao** [comment]
resubmitted by PR2959

**Pull request closed without merging**